usb/device/targets/TARGET_STM/USBPhy_STM32.cpp

```diff
@@ -0,0 +1,522 @@
+{
+    /*  check that endpoint maximum size is not exceeding TX fifo */
+    MBED_ASSERT(hpcd.IN_ep[0].maxpacket >= size);
+    endpoint_write(EP0IN, buffer, size);
+}
+
+void USBPhyHw::ep0_stall()
+{
+    endpoint_stall(EP0IN);
+    endpoint_stall(EP0OUT);
+}
+
+bool USBPhyHw::endpoint_add(usb_ep_t endpoint, uint32_t max_packet, usb_ep_type_t type)
+{
+    uint32_t len;
+    if (max_packet > MAXTRANSFER_SIZE) return false;
+    if (endpoint & 0x80) {
+        HAL_PCDEx_SetTxFiFo(&hpcd, endpoint & 0x7f, (max_packet / 4) + 1);
+        len = HAL_PCDEx_GetTxFiFo(&hpcd,endpoint & 0x7f);
+        MBED_ASSERT(len >= max_packet);
+    }
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Open(&hpcd, endpoint, max_packet, type);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    return (ret == HAL_OK) ? true:false;
+}
+
+void USBPhyHw::endpoint_remove(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret = HAL_PCD_EP_Close(&hpcd, endpoint);
+    MBED_ASSERT(ret == HAL_OK);
+}
+
+void USBPhyHw::endpoint_stall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_SetStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+void USBPhyHw::endpoint_unstall(usb_ep_t endpoint)
+{
+    HAL_StatusTypeDef ret;
+    ret = HAL_PCD_EP_ClrStall(&hpcd, endpoint);
+    MBED_ASSERT(ret!=HAL_BUSY);
+}
+
+bool USBPhyHw::endpoint_read(usb_ep_t endpoint, uint8_t *data, uint32_t size)
+{
+    _read_buf[EP_TO_LOG(endpoint)] = data;
+    _read_size[EP_TO_LOG(endpoint)] = size;
+
+    uint8_t* pBuf = (uint8_t *)pBufRx; //TODO - this buffer shouldn't be shared for multiple endpoints
+    HAL_StatusTypeDef ret;
+    // clean reception end flag before requesting reception
+    ret = HAL_PCD_EP_Receive(&hpcd, endpoint, pBuf, size);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    return true;
+}
+
+uint32_t USBPhyHw::endpoint_read_result(usb_ep_t endpoint)
+{
+    uint8_t *buffer =_read_buf[EP_TO_LOG(endpoint)];
+    uint8_t max_size = _read_size[EP_TO_LOG(endpoint)];
+
+    if (epComplete[EP_TO_IDX(endpoint)]==0) {
+        /*  no reception possible !!! */
+        return 0;
+    } else if ((epComplete[EP_TO_IDX(endpoint)]!=1)) {
+        return 0;
+    }
+    uint8_t  *buff = (uint8_t *)pBufRx;
+    uint32_t length = (uint32_t) HAL_PCD_EP_GetRxCount(&hpcd, endpoint);
+    memcpy(buffer, buff, length > max_size ? max_size : length);
+    epComplete[EP_TO_IDX(endpoint)]= 0;
+    return length;
+}
+
+bool USBPhyHw::endpoint_write(usb_ep_t endpoint, uint8_t *data, uint32_t size)
+{
+    HAL_StatusTypeDef ret;
+    // clean transmission end flag before requesting transmission
+    epComplete[EP_TO_IDX(endpoint)] = 2;
+    ret = HAL_PCD_EP_Transmit(&hpcd, endpoint, data, size);
+    MBED_ASSERT(ret!=HAL_BUSY);
+    // update the status
+    if (ret != HAL_OK) return false;
+    // fix me return is too simple
+    return true;
+}
+
+void USBPhyHw::endpoint_abort(usb_ep_t endpoint)
+{
+    // TODO - stop the current transfer on this endpoint and don't call the IN or OUT callback
```

Comment:
I didn't see a function in the ST cube library. Does it have this functionality, and if not, is there a way to add it? This will be required if the buffer passed into endpoint_read can be used directly.

Reply:
Agree so the buffer read must be removed from Hal internal descriptor to avoid the copy on a memory area no more allocated for usb buffer reception , and the current transaction that has been programmed but not yet performed must be stopped, to allow the queuing of a new read request on same endpoint.  There is no function implemented this abort, a  new function HAL abort should disable the endpoint.  The function For F2,  will be something like the code in following commit https://github.com/jamike/mbed/commit/005e6b379acd043cb3e1699531c0045c389a2928
, this needs to be tested.

Reply:
@jamike after fixes the return code check in endpoint_abort I was able to use your commit and pass testing. Endpoint abort tests aren't complete though so it wasn't exhaustively tested.

Would it be possible to get an endpoint abort function for all ST devices in a future release of the ST Cube library?